route/bgmetadata.go

```diff
@@ -0,0 +1,153 @@
+package route
+
+import (
+	"context"
+	"fmt"
+	"hash/crc32"
+	"sync"
+	"time"
+
+	"github.com/graphite-ng/carbon-relay-ng/encoding"
+	"github.com/graphite-ng/carbon-relay-ng/matcher"
+	"github.com/willf/bloom"
+)
+
+type shard struct {
+	filter  bloom.BloomFilter
+	channel chan []byte
+	lock    sync.RWMutex
+}
+
+// BloomFilterConfig contains filter size and false positive chance for all bloom filters
+type BloomFilterConfig struct {
+	N              uint
+	P              float64
+	ShardingFactor int
+	ClearInterval  time.Duration
+	ClearWait      time.Duration
+}
+
+// BgMetadata contains data required to start, stop and reset a metric metadata producer.
+type BgMetadata struct {
+	baseRoute
+	shards []shard
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+	bfCfg  BloomFilterConfig
+}
+
+// NewBgMetadataRoute creates BgMetadata, starts sharding and filtering incoming metrics.
+// Runs a goroutines for each shard to handle incoming metrics and periodic cleanup of bloom filters
+func NewBgMetadataRoute(key, prefix, sub, regex string, bfCfg *BloomFilterConfig) (*BgMetadata, error) {
+	m := BgMetadata{
+		baseRoute: *newBaseRoute(key, "bg_metadata"),
+		shards:    make([]shard, bfCfg.ShardingFactor),
+		bfCfg:     *bfCfg,
+	}
+
+	if bfCfg.ClearWait != 0 {
+		m.bfCfg.ClearWait = bfCfg.ClearWait
+	} else {
+		m.bfCfg.ClearWait = bfCfg.ClearInterval / time.Duration(bfCfg.ShardingFactor)
+		m.logger.Info(fmt.Sprintf("setting clear_wait value to %s", m.bfCfg.ClearWait.String()))
+	}
+
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+
+	// init every shard with filter and channel
+	for shardNum := 0; shardNum < bfCfg.ShardingFactor; shardNum++ {
+		m.shards[shardNum] = shard{
+			filter:  *bloom.NewWithEstimates(bfCfg.N, bfCfg.P),
+			channel: make(chan []byte, 500),
+		}
+		go m.handleMetric(shardNum)
+	}
+	m.logger.Info("all metric handlers started")
+
+	go m.clearBloomFilter()
+
+	// matcher required to initialise route.Config for routing table, othewise it will panic
+	mt, err := matcher.New(prefix, sub, regex)
+	if err != nil {
+		return nil, err
+	}
+	m.config.Store(baseConfig{*mt, nil})
+
+	return &m, nil
+}
+
+func (m *BgMetadata) handleMetric(shardNum int) {
+	m.wg.Add(1)
+	defer m.wg.Done()
```

Comment:
fetch the correct shard only once instead of duplicating accessing code

Reply:
it's not relevant to this code block anymore. i moved the logic to Dispatch() and i'm not duplicating access code there.